activerecord/test/cases/enum_test.rb

```diff
@@ -508,4 +508,24 @@ def self.name; "Book"; end
   test "data type of Enum type" do
     assert_equal :integer, Book.type_for_attribute("status").type
   end
+
+  test "enum with string attribute defined as hash" do
+    klass = Class.new(ActiveRecord::Base) do
+      self.table_name = "books"
+      enum cover: { hardcover: :hardcover, paperback: :paperback }
+    end
+
+    book = klass.hardcover.create!
+    assert_equal "hardcover", klass.last.cover
+  end
+
+  test "enum with string attribute defined as array of symbols" do
+    klass = Class.new(ActiveRecord::Base) do
+      self.table_name = "books"
+      enum cover: %i[hardcover paperback]
```

Comment:
Does it means that existing apps need to change the code if want to keep the original behavior?

Reply:
So it won't break the original behavior.

Reply:
This PR would affect created data for enum on string column.
So existing apps which use enum on string need to change the code if want to keep the original behavior.

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
"activerecord_unittest2")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :items, force: true do |t|
    t.string :code
  end
end

class Item < ActiveRecord::Base
  enum code: %i(code_a code_b code_c)
end

class BugTest < Minitest::Test
  def test_enum
    item = Item.create!(code: "code_a")
    assert_equal item, Item.find_by(code: 0)
  end
end
```

Reply:
Yes. You're right. The code you've written above will be broken when using the new enum-to-string mapping.

But take a look, what kind of records would it actually store to the the database with the original implementation:

* `"0"` for :code_a
* `"1"` for :code_b
* `"2"` for :code_c

So. The numeric indices will be converted to string values and stored as characters. Who would ever design his code in a such way?

Reply:
If you do really worry about this implementation, I can add an option `string_enum: true` (or whatever you like) wich will turn the new behaviour on only for particular cases.
The only reason I haven't done it before is that I wanted to keep this "magical" guessing of ROR.

Reply:
Nice catch, @kamipo. Maybe we need to deprecate the existing behaviour first:

> The enum `code` on the model `Item` is backed by a string column. Its value is currently stored in the database as an integer, but in Rails 6.1 the label will be stored directly instead.
> To preserve the current behaviour, provide an explicit mapping of labels to database values:
>     enum code: { code_a: "0", code_b: "1", code_c: "2" }

Reply:
I'm +1 on this. I would prefer that we deprecate the existing behaviour, rather than adding any extra options.